### DISCUSSION TITLE
Migration to independent platform for provision of source code

### DISCUSSION THREAD

**Discussion by linux-lukas** [containers/podman#15045] [General]
### A. Problem / Goal
Since the purchase of GitHub by Microsoft in 2018, a dependence on the BigTech corporation can no longer be denied.

On the one hand, I can understand why GitHub was chosen as the platform for making source code available: "Everyone is here".

On the other hand, I see the danger of a vendor-lockin effect and that open source projects become centrally dependent on Microsoft. In my eyes, this is very dangerous for free and open source software and hardware projects.

In the medium and long term, the goal would be to become independent of GitHub and thus of Microsoft. The Gitea-based Codeberg project of Codeberg e. V. in Berlin would be a good choice here.

There are also (legal) problems with compliance with the licence of GitHub functions, such as the co-pilot.

### B. Solution
My considered solution to the problem described in A. would be the following:

A user of this open source project creates a user account on https://codeberg.org/
If necessary: This user creates an organisation for the project.
A "personal access token" is created on the GitHub account, which has appropriate rights to the organisation repositories, using the developer options in the settings.
all repositories would be migrated with this access token into the ownership of the organisation created in step two.
Regarding step four, there is an entry in the documentation of Codeberg: https://docs.codeberg.org/advanced/migrating-repos/

### C. Alternatives
A possible alternative would be to perform the first three steps as described in B. A possible alternative would be to perform the first three steps as described in B., and modify the fourth step to include a mirror of GitHub. So that all issues and such that would be created in the GitHub repository would be transferred to the Codeberg repository.

### D. Responsibilities
I would see the responsibility in the owners of the repository and, if necessary, additional project participants.

### E. Other
Basically, a look at the documentation of Codeberg is not unwise: https://docs.codeberg.org/

Should it be necessary to manage repositories in organisations, this is also possible under Codeberg, see: https://docs.codeberg.org/collaborating/create-organization/

Regarding licensing there is a page in the documentation of Codeberg: https://docs.codeberg.org/getting-started/licensing/

Funfact 1: The Pixelfed maintainer has written on Mastodon that he is considering migrating the Pixelfed project away from GitHub, to Codeberg: https://mastodon.social/@dansup/10869007146309141

Funfact 2: Aral had migrated - also on Mastodon - to Codeberg for his GitHub projects as well as those of his Foundation, which he organised together with another person, Small Technology, or is still migrating: https://mastodon.ar.al/@aral/108651118032076133

**Comment by vrothberg** [comment]
Thanks for reaching out, @Lukas2112!

I try to see things pragmatically. Personally, I have absolutely no problem with GitHub's relationship/association with Microsoft or Microsoft as such. I don't think that vendor lock-in is a real problem since there are good alternatives to GitHub such as Gitlab. In my view, the main pain points of such a migration are the (immense) costs (in working hours) which would apply to any other platform as well.

**Comment by rhatdan** [comment]
I would add the network effect.  The contributors exist in github.com, changing forces new contributors to setup new secondary accounts on the new sight.  This I believe would cut down on your chance to contribute.

**Reply by linux-lukas** [reply]
Yes, the outreach thing is of course true on the one hand, but nevertheless it should be hoped that GitHub will move away from its centralised approach and towards a decentralised approach. There is an effort within the Gitea project to match the Fediverse with its decentralised approach to software development and to enable federation between Gitea instances - if Gitea supports this, it would no longer be such a priority where the code is located, whether on GitHub - but it would still make sense to change in the long term.

**Comment by rhatdan** [comment]
Microsoft over time has become much more friendly to OpenSource, perhaps a wolf in sheeps clothing but I don't think so.  Microsofts goal is to get people to use their cloud Azure, and doing things Anti-OpenSource would hurt that goal.

**Comment by Luap99** [comment]
I personally would like to use a platform which is actually open source.

That said I understand why GitHub is used by so many project and I don't think the argument vendor lock in is really good.
If you want to convince people to switch you have to offer them something better, i.e. will another platform increase our productivity? I don't think so.

I agree with @vrothberg. The cost of switching are extremely high. Porting the git repo is easy, porting issues/pull request and most important our complex CI system is not.

**Reply by linux-lukas** [reply]
I agree with you that it would not be a small effort. A complete migration that ensures at least equivalent ease of use. Gitea, which Codeberg is based on, is well on its way to achieving this.

**Comment by mheon** [comment]
Gitlab seems to support repository mirroring, so it should be possible to mirror our repository to Gitlab. I think a read-only mirror might not be a bad idea, partially as a hedge against a future need to switch to away from Github, partially as a mirror for when Github goes down (which seems to be often enough of late to justify it).

**Reply by linux-lukas** [reply]
In principle, it would be possible, but I would like to note that Gitlab uses Cloudflare to host its instance and that this is a reason for users not to participate in the project because, as an EU citizen, there are well-founded concerns about US companies in the technology sector because of legal situations in the US.
